src/core/toga/evented.py

```diff
@@ -0,0 +1,164 @@
+    ...
+    ...     def __init__(self, **kwargs):
+    ...         super().__init__(**kwargs)
+    ...         # This is typically done via the `factory` argument
+    ...         self._impl = AClickableWidgetBackend(interface=self)
+    ...
+    >>> class AClickableWidgetBackend:
+    ...     def __init__(self, interface):
+    ...         self.interface = interface
+    ...
+    ...     def emulate_click(self, which_button):
+    ...         self.interface.raise_event('on_click', which_button)
+    ...
+    >>> def click_handler(widget, which_button):
+    ...     print(f"I was clicked with {which_button} mouse button!")
+    ...
+    >>> widget = AClickableWidget(on_click=click_handler)
+    >>> widget._impl.emulate_click("Left")
+    I was clicked with Left mouse button!
+    >>> widget._impl.emulate_click("Right")
+    I was clicked with Right mouse button!
+
+Note: Typically only top-level classes such as `Widget` or `Window` should
+inherit directly from `Evented`, others should inherit from those classes
+instead.
+"""
+from toga.handlers import wrapped_handler
+
+
+class Evented:
+    """Base class for classes that can have event callbacks
+
+    All keyword arguments passed to this class become attempts at setting event
+    callbacks.
+
+    Objects of this class are not meant to be used directly. Instead, other
+    classes that wish to be able to call event callbacks should inherit from it
+    and then define some events using the Event descriptor class.
+    """
+    def __init__(self, **kwargs):
+        for arg, callback in kwargs.items():
+            attr = self._get_event_descriptor(arg)
+            if attr is None or not isinstance(attr, Event):
+                continue
+            attr.__set__(self, callback)
+
+    @property
+    def _impl(self):
+        return self.__impl
+
+    @_impl.setter
+    def _impl(self, impl):
+        """We intercept the setting of the _impl attribute so we could notify
+        the implementation object about callbacks that had been set before
+        that object was created (E.g. via the constructor).
+        """
+        self.__impl = impl
+        # When we get a new implementation backend, let it know about all the
+        # callbacks we have
+        for klass in type(self).mro():
+            for attr in klass.__dict__.values():
+                if not isinstance(attr, Event):
+                    continue
+                attr._notify_impl(self)
+
+    def raise_event(self, event, *args, **kwargs):
+        """Call the callback associated with an event, if any
+
+        Args:
+            event (str): The name of the event to raise callbacks for
+
+        All other positional and keyword arguments are passed to the callback
+        function. If a bad event string is passed, UndefinedEventRaised
+        exception is raised.
+        """
+        descr = self._get_event_descriptor(event)
+        if descr is None:
+            raise UndefinedEventRaised
```

Comment:
Although this is level, it's generally better style to raise an exception instance, rather than a bare class.

Reply:
whoops, I did not really intend to do that, will fix